File: primalTestDotNet/AI/States/UnwinnableState.cs
using PrimalTestDotNet.Entities;

namespace PrimalTestDotNet.AI.States;

public class UnwinnableState(Level.Level level) : StateBase(level), IState
{
    public void Update(Hero hero, IEnumerable<IGameObject> gameObjects)
    {
        var distanceData = CalculateDistanceData(hero.Position, []);
        // forfeit when unwinnable
        // try to exit
        var exit = gameObjects.FirstOrDefault(o => o is Exit);
        var pathToExit = GetPath(hero.Position, exit!.Position, distanceData);
        var node = pathToExit.Last();
        var direction = node - hero.Position;

        if (direction.Length > 1)
        {
            // If leaving without the treasure also fails => game over 
            _level.OnForfeit();
            return;
        }

        hero.Move(direction);
    }
}

File: PrimalTestCpp/UnwinnableState.cpp
#include "UnwinnableState.h"
#include "Hero.h"
#include "Exit.h"

UnwinnableState::UnwinnableState(AIStateMachine& context, Level& level) : StateBase(level), _context(context)
{
}

void UnwinnableState::update(Hero& hero, std::vector<std::shared_ptr<IGameObject>>& gameObjects)
{
	auto distanceData = calculateDistanceData(hero.getPosition(), {});
	// forfeit when unwinnable
	// try to exit
	auto exits = findAllObjectsOfType<Exit>(gameObjects);
	auto& exit = exits.front();
	auto pathToExit = getPath(hero.getPosition(), exit->getPosition(), distanceData);
	auto& node = pathToExit.back();
	auto direction = node - hero.getPosition();

	if (direction.getLength() > 1) {
		// cant leave for some reason, exit is closed
		_level.onForfeit();
		return;
	}

	// move to exit
	hero.move(direction);
}
